sshd-sftp/src/main/java/org/apache/sshd/client/subsystem/sftp/impl/SftpInputStreamAsync.java

```diff
@@ -0,0 +1,312 @@
+        }
+        long orgOffset = clientOffset;
+        while (!eofIndicator) {
+            if (hasNoData()) {
+                fillData();
+                if (eofIndicator && hasNoData()) {
+                    break;
+                }
+                sendRequests();
+            } else {
+                int nb = buffer.available();
+                out.write(buffer.array(), buffer.rpos(), nb);
+                buffer.rpos(buffer.rpos() + nb);
+                clientOffset += nb;
+            }
+        }
+        return clientOffset - orgOffset;
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
+        if (!isOpen()) {
+            throw new IOException("skip(" + getPath() + ") stream closed");
+        }
+        if (clientOffset == 0 && pendingReads.isEmpty()) {
+            clientOffset = n;
+            return n;
+        }
+        return super.skip(n);
+    }
+
+    boolean hasNoData() {
+        return buffer == null || buffer.available() == 0;
+    }
+
+    void sendRequests() throws IOException {
+        if (!eofIndicator) {
+            long windowSize = client.getChannel().getLocalWindow().getMaxSize();
+            while (pendingReads.size() < (int) (windowSize / bufferSize) && requestOffset < fileSize + bufferSize
+                    || pendingReads.isEmpty()) {
+                Buffer buf = client.getSession().createBuffer(SshConstants.SSH_MSG_CHANNEL_DATA,
+                        23 /* sftp packet */ + 16 + handle.getIdentifier().length);
+                buf.rpos(23);
+                buf.wpos(23);
+                buf.putBytes(handle.getIdentifier());
+                buf.putLong(requestOffset);
+                buf.putInt(bufferSize);
+                int reqId = client.send(SftpConstants.SSH_FXP_READ, buf);
+                pendingReads.add(new Ack(reqId, requestOffset, bufferSize));
+                requestOffset += bufferSize;
+            }
+        }
+    }
+
+    void fillData() throws IOException {
+        Ack ack = pendingReads.pollFirst();
+        if (ack != null) {
+            pollBuffer(ack);
+            if (!eofIndicator && clientOffset < ack.offset) {
+                // we are actually missing some data
+                // so request is synchronously
+                byte[] data = new byte[(int) (ack.offset - clientOffset + buffer.available())];
+                int cur = 0;
+                int nb = (int) (ack.offset - clientOffset);
+                AtomicReference<Boolean> eof = new AtomicReference<>();
+                while (cur < nb) {
+                    int dlen = client.read(handle, clientOffset, data, cur, nb - cur, eof);
+                    eofIndicator = dlen < 0 || eof.get() != null && eof.get();
+                    cur += dlen;
+                }
+                buffer.getRawBytes(data, nb, buffer.available());
+                buffer = new ByteArrayBuffer(data);
+            }
+        }
+    }
+
+    void pollBuffer(Ack ack) throws IOException {
```

Comment:
Let's make it protected in accordance with our policy to allow maximum flexibility to users who wish to modify/enhance the behavior of the relevant class